### PR TITLE
Adding Trader Workstation option, SSH authentication with private key file and automatic option for ActiveX and Socket Clients

### PR DESCRIPTION
**Pull request by nalon99**
In addition to this awesome repo I would like to give my little contribution by adding some other features:

1. Trader Workstation other than Gateway
2. ssh authentication with a private key
3. Enabling “ActiveX and Socket Clients” option without manual user intervention

- to configure the public key in the SSH server this must be set in a new environment variable: SSH_PUBLIC_KEY
- added another environment variable GATEWAY_OR_TWS where, if set to **tws** starts Trader Workstation or starts Gateway if set to **gateway**
- added AUTO_LOGOFF_TIME which sets when logging off TWS or Gateway application at a specific time
- added SAVE_TWS_SETTINGS used to specify the time to save settings to Interactive Brokers (look at IBC documentation for more info)

Here is an example on how to define a .env including these additional variables:
```
GATEWAY_OR_TWS=tws # or gateway
AUTO_LOGOFF_TIME="10:30 PM" # UTC time
SAVE_TWS_SETTINGS="21:30" # UTC time
SSH_PUBLIC_KEY="ssh-rsa xxxxxxxxxx"
```

### CODE REVIEW

**Comment by gnzsnz** [comment]
thanks @nalon99  for the PR.

Please give me a few days to review it.

Please note that there is already support for SSH. With an ssh-client (as it is in the latest and stable releases) we can tunnel gateway and VNC ports. I didn't want to include an SSH-server as this is a docker [anti-patern](https://jpetazzo.github.io/2014/06/23/docker-ssh-considered-evil/). So it's unlikely that I will include the ssh-server.

**Comment by nalon99** [comment]
No hurry, take your time for the review.
Anyway, it might be useful to keep the option to run TWS instead of the Gateway, and in particular to automatically enable ActiveX and Socket Clients which is normally disabled on TWS and requires manually intervention.
________________________________
Da: gnzsnz ***@***.***>
Inviato: sabato 18 novembre 2023 10:38
A: gnzsnz/ib-gateway-docker ***@***.***>
Cc: Andrea Nalon ***@***.***>; Mention ***@***.***>
Oggetto: Re: [gnzsnz/ib-gateway-docker] Adding Trader Workstation option, SSH authentication with private key file and automatic option for ActiveX and Socket Clients (PR #43)


thanks @nalon99<https://github.com/nalon99> for the PR.

Please give me a few days to review it.

Please note that there is already support for SSH. With an ssh-client (as it is in the latest and stable releases) we can tunnel gateway and VNC ports. I didn't want to include an SSH-server as this is a docker anti-patern<https://jpetazzo.github.io/2014/06/23/docker-ssh-considered-evil/>. So it's unlikely that I will include the ssh-server.

**Comment by gnzsnz** [comment]
> No hurry, take your time for the review. Anyway, it might be useful to keep the option to run TWS instead of the Gateway, and in particular to automatically enable ActiveX and Socket Clients which is normally disabled on TWS and requires manually intervention.

I went though your PR, and I would like to propose a way forward. Basically there are things that I'm OK with, so I would propose that we include them in the code base. And things that I think need more exploration.

**Things that can be committed**

- `AUTO_LOGOFF_TIME`, please see my comment on the review
- `BYPASS_WARNING`, please see my comment on the review

Please re-sync your fork. It's no in sync with master

**Things that need more exploration**

- `SAVE_TWS_SETTINGS` I don't see the use case for this one. I'm not against it, as it's an IBC parameter. But I don't see why it's needed in a container, neither for ib-gateway or even TWS. Please elaborate as described on the review.
- `GATEWAY_OR_TWS` I didn't know that this was possible. I did a test and indeed it works, but it needs a few changes. in my tests this needed these libraries `libgtk-3-0 libasound2 libnss3 libgbm1 libnspr4` on the `Dockerfile`. And although it works, I fell this is very basic. What is your rational to use TWS in a docker headless image?

I'm open to the idea, but I would like something that covers all user cases. In special I think it needs a web browser, otherwise there are many things that are not available, and an spreadsheet. A web browser like [Epiphany](http://projects.gnome.org/epiphany/) and [gnumeric](http://www.gnumeric.org/) could be a basic option. But I don't think that "basic" will do the trick.
In addition, we would need a way to pass to `Xvfb` the resolution. the default `1024x768` is not good enough for TWS.

Could you please elaborate on how you expect TWS to work in the container? and how it should be used.

- SSH server - as mentioned I'm not really open to the idea of having an SSH server. The latest `stable`and `latest` releases include an SSH client. So SSH tunnels are available. Please feel free to justify why you think it's needed.
-

**Review by gnzsnz — CHANGES_REQUESTED** [review]
Please see my comments

File: docker-compose.yml
Comment (nalon99) [reply docker-compose.yml:20]: Sure, agree.

File: docker-compose.yml
Comment (nalon99) [reply docker-compose.yml:21]: The idea of implementing TWS in a container, instead of Gateway, is to allow the owner of Interactive Brokers account (for example your Client) to connect to. Let's imagine you created a trading bot application and you want to connect it to IB account of your Client. Your trading bot is running somehwere in the cloud and not in the same instance where TWS is running (because TWS is running on your Client cloud and you own the intellectual property of the trading bot). But how can your Client access to TWS and see the list of trades, or monitor what's going on? he can't open TWS on his computer since it will disconnect the TWS running in the Cloud (and your trading bot too). So he needs to connect remotely to the TWS in the cloud.
One problem with TWS is that there are settings which are not controlled by IBC and you need to apply them manually. Therefore, they need to be saved on IB servers in the Client profile.
From my staindpoint, the container will still be stateless because these settings are saved outside the container and into the Client profile.
Of course, this is my proposal but the final decision is up to you :-)

File: docker-compose.yml
Comment (nalon99) [reply docker-compose.yml]: I changed a little bit run.sh... GATEWAY_OR_TWS now accepts: tws | gateway

File: docker-compose.yml
Comment (nalon99) [inline docker-compose.yml:28]: I don't see many options to protect a container in the cloud running TWS with api sockets open on a public internet network, than shielding these ports (as well as VNC port) with SSH. The client will use its private key to authenticat to the server running TWS (or Gateway).
Another tool to protect it, could be fail2Ban

File: image-files/scripts/run.sh
Comment (nalon99) [inline image-files/scripts/run.sh:14]: is it more clear to pass a parameter value tws or gateway instead of a command -g?

File: stable/Dockerfile
Comment (nalon99) [inline stable/Dockerfile:64]: to run TWS in the container libgtk 3.0 is the only necessary library to add

**Review by nalon99 — COMMENTED** [review]
please have a look at my comments, thanks.

File: docker-compose.yml
Comment (gnzsnz) [reply docker-compose.yml:28]: there is already SSH support on the container. see --> https://github.com/gnzsnz/ib-gateway-docker#ssh-tunnel

SSH support is implemented with ssh-client, which means that the container can connect an ssh server. not the other way around. this avoid implementing an ssh-server,  which is an anti-patern. please remove all changes related to ssh server.

this one is a "no"

File: docker-compose.yml
Comment (gnzsnz) [reply docker-compose.yml]: I see your point, and I like the idea of having TWS as an option. Specially because we can use the same "IB Gateway" download to run ib-gateway or TWS by just switching the parameter on IBC start-up script.

As I say on my previous [reply](https://github.com/gnzsnz/ib-gateway-docker/pull/43#issuecomment-1817528069) there are thinks that can be committed and things that need further analysis. I think we can pass `TWS_SETTINGS_PATH` into things that can be committed. But please update it to `${SAVE_TWS_SETTINGS:-}`

File: stable/Dockerfile
Comment (gnzsnz) [reply stable/Dockerfile:64]: libgtk 3.0 was not the only thing i needed in my test. that's strange.

in any case, I do like the idea of having TWS, but to be honest, I don't think this is a good 'user experience'. there is no desktop, windows manager, no browser, it rather limited.

Please take these lines our of the PR. I think this needs more analysis. I'm thinking on implementing the change using a different base image, something like [docker-rdesktop](https://github.com/linuxserver/docker-rdesktop). we can run IBC on that image with a fully functional desktop environment.

One important point, please take out all changes on folders stable and latest. Changes should be on the dockerfile.template or on image-script/* forlder.

File: image-files/config/ibc/config.ini.tmpl
Comment (gnzsnz) [inline image-files/config/ibc/config.ini.tmpl:1]: this file is OK to go.

File: image-files/config/ibc/config.ini.tmpl
Comment (gnzsnz) [inline image-files/config/ibc/config.ini.tmpl:791]: sorry, this file is OK to go, except for this line. as this is related to TWS. please take it out for the time being. as this needs to go together witht the telnet piece

File: stable/Dockerfile
Comment (gnzsnz) [reply stable/Dockerfile:64]: @nalon99 , could you please check #45 ? I have explored the path of [docker-rdesktop](https://github.com/linuxserver/docker-rdesktop) and i think it looks promising. I would appreciate your feedback.

File: docker-compose.yml
Comment (nalon99) [reply docker-compose.yml:28]: OK. I'm creating a local branch just for me keeping ssh server because in my flow, it's the container running the trading bot logic which connects to the instance running TWS.

File: stable/Dockerfile
Comment (nalon99) [reply stable/Dockerfile:64]: rdp protocol is faster and more secure than standard vnc protocol and rdp clients are available also for Linux and Mac os. It would be a good chance to use it.

**Comment by nalon99** [comment]
please, remove script files from stable directory before merging

File: docker-compose.yml
Comment (gnzsnz) [reply docker-compose.yml]: Hi, you can implement the [tunnel](https://github.com/gnzsnz/ib-gateway-docker#ssh-tunnel) with minimal changes on your current logic.

the idea behind containers is that they should do one thing, an SSH server requires setup, maintenance and is a risk. that's why it should be in a separate container, I have build a container for [SSH](https://github.com/gnzsnz/docker-bastion) to address this. there are other containers available, so there is choice.

based on your case, you can 

1. set-up an SSH container(most of the work is to setup user/host Keys, using [SSH certificates][1] simplifies this.)
2. connect your trading bot to the SSH container, this is a minimal change. just change the host you connect to.
3. connect ib-gateway to SSH container, this is already supported

your ib-gatewa is secured, your trading bot is secured, and you use a security hardened SSH container. Tunnels eventually get disconnected, but is the same than with your current setup. ib-gateway has a logic to restart tunnels so this is mitigated.

i'm happy to help on this one.

for reference: https://goteleport.com/blog/how-to-ssh-properly/


[1]: https://gist.github.com/gnzsnz/3fcfa368fa3bc181c8bd69da96037b2c

File: stable/Dockerfile
Comment (gnzsnz) [reply stable/Dockerfile]: > rdp protocol is faster and more secure than standard vnc protocol and rdp clients are available also for Linux and Mac os. It would be a good chance to use it.

I've been pushing commits into #45  branch and it's almost ready. if you can please test it, i would appreciate your tests and feedback. the code is ready for building a local image for "latest", basically just use `tws-docker-compose.yml` file

by the way, I'm able to run tws with IBC and "activeX and socker client" are enabled by default, I don't need to do anything to have the API working

the way I see this is as a different image, one for ibgateway and one for tws. with a common logic for IBC, socat and ssh tunnels.

**Pull request closed without merging**
